### PR TITLE
[Mobile Payments] Pass gateway ID to existing IPP onboarding Tracks events

### LINKED ISSUE #10596: [Mobile Payments] Pass `gatewayID` to existing IPP onboarding track events
Context: pdfdoF-3HA#comment-4936-p2

We do not pass the _payment gateway ID_ as an explicit property to some IPP onboarding events, and rely on the event identifier itself to know if is WCPay or others. Ideally, we want to pass this explicitly into the event property list, for example as we do in `_card_present_collect_payment_success`

We should also leave a note in the track validation to clarify this (ref pdfdoF-3HA#comment-4945-p2 )

### PR DESCRIPTION
**Pull request by jaclync**
Closes: #10596 

## Why

Based on the issue description:

We currently do not pass the _payment gateway ID_ as an explicit property to some IPP onboarding events, and rely on the event identifier itself to know if is WCPay or others. Ideally, we want to pass this explicitly into the event property list, for example as we do in `_card_present_collect_payment_success`. This PR adds `plugin_slug` property to all 6 `card_present_onboarding_*` events.

## How

The 6 `card_present_onboarding_*` events were updated to include a new `plugin_slug` (gateway ID) property. `gatewayID: String?` extension was added to `CardPresentPaymentOnboardingState` for easier access in a few events. The changes are a bit scattered as the plugin needs to be DI'ed or inferred by looking at the code. I also deleted `CardPresentPaymentOnboardingState.enabled` case because it's not used anymore.

## Testing instructions

Hand-testing is optional, as it can be time-consuming to test all scenarios. There are also a lot of scenarios that could be hard to test like for a few error scenarios. I will make sure to ensure the event property `plugin_slug` is tracked for all 6 events.

- [x] @jaclync updates event registration in Tracks

---

- [x] `card_present_onboarding_step_skipped`
  - Find a store with the payment plugin all set up
  - In wp-admin > payment methods (`wp-admin/admin.php?page=wc-settings&tab=checkout`), disable `Cash on delivery` payment method and tap `Save changes`
  - In the app, switch to the store
  - Go to the Menu tab > Payments
  - Tap `Continue setup` --> should see a screen like in [this issue](https://github.com/woocommerce/woocommerce-ios/issues/7525))
  - Tap `Skip for now`
- [x] `card_present_onboarding_cta_tapped`, `card_present_onboarding_cta_failed`
  - Switch to a store with a payment plugin but not set up yet
  - Go to the Menu tab > Payments
  - Tap `Continue setup` 
  - Tap `Finish Setup in Store Admin` --> both events should be tracked with `plugin_slug`
- [x] `card_present_onboarding_learn_more_tapped`
  - Switch to a store with a payment plugin but not set up yet
  - Go to the Menu tab > Payments
  - Tap `Continue setup`
  - Tap `Learn more` --> the event should be tracked with `plugin_slug`
- [x] `card_present_onboarding_not_completed`
  - Switch to a store with a payment plugin but not set up yet
  - Go to the Menu tab > Payments
  - Tap `Continue setup`
  - Navigate back to the payments screen --> the event should be tracked with `plugin_slug`
- [x] `card_present_onboarding_completed`
  - Switch to a store with payments set up already
  - Go to the Menu tab > Payments --> the event should be tracked with `plugin_slug`

---
- [x] I have considered if this change warrants user-facing release notes and have added them to `RELEASE-NOTES.txt` if necessary.

### CODE REVIEW

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift
Comment (joshheald) [inline Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:12]: I don't think it's safe to delete `enabled`... do you have any context on why we can remove it now?

Often WooPayments returns `.enabled` instead of `.completed`. This happens when Stripe report `has_future_requirements: true` for the connected account. That means "this user can take payments, but when they hit some threshold, we'll need more KYC information from them". For our purposes, we treat this as equivalent to `completed`.

`enabled` was added because merchants with `enabled` would fall into the default case, which was a catch-all onboarding error. 

Removing it adds a big risk that merchants who have (essentially) completed onboarding will be shown an error and unable to take payments.

**Review by joshheald — COMMENTED** [review]
Still reviewing, but this seemed worth mentioning straight away

File: WooCommerce/Classes/Analytics/WooAnalyticsEvent.swift
Comment (bozidarsevo) [inline WooCommerce/Classes/Analytics/WooAnalyticsEvent.swift]: maybe we can rename this to something like `safeGatewayID(forGatewayID gatewayID` or something like that?
Now it says gatewayId for gatewayId and it feels a bit weird :D

**Review by joshheald — COMMENTED** [review]
Thanks for adding these 😊  

The analytics work fine with my testing. Would approve, if it weren't for the previously mentioned issue around `enabled`, which needs more discussion/to be put back.

File: Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift
Comment (jaclync) [reply Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:12]: Thanks for catching and clarifying this, I meant to add a comment to confirm this removal but it slipped my mind at the time. Do you know where the state can be set to `enabled`? I didn't know about the context for this state, for the goal of this PR I first tried adding an associated value for the plugin to the `enabled` case, but I looked at `CardPresentPaymentsOnboardingUseCase` and a few other use cases and couldn't find where the state can be set to `enabled` (I tried updating `CardPresentPaymentsOnboardingUseCaseProtocol.state` to be readonly and `CardPresentPaymentsOnboardingUseCase.state` to be `private(set)` and `CardPresentPaymentsOnboardingUseCase` seems to be the only class that sets the state). Since `CardPresentPaymentOnboardingState` is an enum defined in Yosemite maintained internally, I thought there has to be some way of setting the state to this case. I'm probably missing something like how this can be set by a value from Stripe, I will find some way to document this since these cases can be removed without failing the build. I'll also need to find a way to get the plugin info for tracking the plugin event prop, it sounds like it might be only relevant in the Stripe plugin.

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift
Comment (joshheald) [reply Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:12]: Hmmm I may be wrong on this precise enum.

`enabled` comes from WooPayments REST API. See `WCPayAccountStatusEnum` for where it comes in over the network. That's never linked to this enum that I can see though. In `CardPresentPaymentsOnboardingUseCase`, we have `accountStatusAllowedForCardPresentPayments()` which checks for `status == .enabled`, which is probably what I'm thinking of.

If you could test it with an account which is in the `enabled` state, we should be able to delete it. Check your account in the Stripe dashboard, to see whether it's `enabled` – it's likely to be.

Comment (iamgabrielma) [reply Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:12]: Just chiming in to add some context:

- p1692608289678179-slack-C02HWCJUFU5 Users reported having troubles setting up IPP, they were landing on an unexpected error when going through the flow. This is when we discovered that the `.enabled` state was added to WCPay but was not handled by the apps.
- p1692780778792139-slack-C025A8VV728 discussion about the specific apps fix

Comment (jaclync) [reply Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:12]: Thanks for sharing the details and context. From my understanding of the changes in the app fix https://github.com/woocommerce/woocommerce-ios/pull/10521 that introduced this `CardPresentPaymentOnboardingState.enabled` case, this case wasn't used in the PR in the first place? 🤔 I didn't see a mention in the PR to use it in a future PR, would be great to confirm with @iamgabrielma.

Comment (iamgabrielma) [reply Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:12]: > this case wasn't used in the PR in the first place? 🤔 I didn't see a mention in the PR to use it in a future PR

I'm not sure to understand this bit, let me know if I miss-understood the question: 

That's right, we did not have that case initially, the account was ready for card present payments as long as the account status coming from Stripe was either `.complete` or `.restrictedSoon`. At some point Stripe introduced `.enabled` as well in this bucket, WCPay made the changes, but the app missed this update.

The app fix on https://github.com/woocommerce/woocommerce-ios/pull/10521 adds the case for `.enabled` since users that had their account in "temporarily good standing" in the web or Stripe were unable to use payments in the app. 

I don't think we can switch the state to enabled for testing (at least not last time I checked), and I'm not sure how troublesome could be setting up a site/account that falls into that specific state. The way we tested it was by intercepting `/wc/v3/payments/accounts*` and changing the status property in the response.

Comment (jaclync) [reply Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:12]: > I'm not sure to understand this bit, let me know if I miss-understood the question:

Sorry if I wasn't clear, I meant that in https://github.com/woocommerce/woocommerce-ios/pull/10521 I didn't see the onboarding state being set to `CardPresentPaymentOnboardingState.enabled` in `CardPresentPaymentsOnboardingUseCase` or anywhere else. The `enabled` state was added to the `WCPayAccountStatusEnum` enum, which is used for a check in `CardPresentPaymentsOnboardingUseCase.accountStatusAllowedForCardPresentPayments`. The test cases were also asserting the onboarding state to `completed`, not `enabled`. Lemme know if I missed something!

**Comment by dangermattic** [comment]
<table>
  <thead>
    <tr>
      <th width="50"></th>
      <th width="100%" data-danger-table="true" data-kind="Warning">
          1 Warning
      </th>
     </tr>
  </thead>
  <tbody>
    <tr>
      <td>:warning:</td>
      <td data-sticky="false">View files have been modified, but no screenshot is included in the pull request. Consider adding some for clarity.</td>
    </tr>
  </tbody>
</table>

<p align="right" data-meta="generated_by_pr-check">
  Generated by :no_entry_sign: <a href="https://danger.systems/">Danger</a>
</p>

File: WooCommerce/Classes/Analytics/WooAnalyticsEvent.swift
Comment (jaclync) [reply WooCommerce/Classes/Analytics/WooAnalyticsEvent.swift]: Good suggestion! Refactored in https://github.com/woocommerce/woocommerce-ios/pull/11690/commits/14237ea1e157b403deb12d61741a277f8163ea7b

**Comment by jaclync** [comment]
thanks for reviewing & testing @joshheald! I responded to the discussion about the `enabled` onboarding state. Also, in https://github.com/woocommerce/woocommerce-ios/pull/11690/commits/66d00ad9094962bb19a211b6f5724686f10bf0ab I updated the onboarding state to be readonly in the `CardPresentPaymentsOnboardingUseCaseProtocol` protocol & implementation so that anyone looking/debugging in the future can just look within the implementation without checking potential external state setters. Lemme know any other feedback/suggestions!

File: Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift
Comment (iamgabrielma) [reply Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:12]: > The enabled state was added to the WCPayAccountStatusEnum enum, which is used for a check in CardPresentPaymentsOnboardingUseCase.accountStatusAllowedForCardPresentPayments. The test cases were also asserting the onboarding state to completed, not enabled.

Gotcha! That's the case, yes. It's only used inside `accountStatusAllowedForCardPresentPayments()` and tests assert for `.completed` state. 

This is outside of this PR, but I think we could benefit from some additional re-thinking/testing here (and perhaps is where you were going with): 

At the moment when we perform the account status checks via `CardPresentPaymentsOnboardingUseCase.accountChecks` and find that internally the `accountStatusAllowedForCardPresentPayments` check passes, we're returning that the `CardPresentPaymentOnboardingState` is `.completed` [if does not return an error](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person%20Payments/CardPresentPaymentsOnboardingUseCase.swift#L407-L417), and while is true that merchants can perform operations at this point as if the onboarding was completed, the account state is not correct at this point, since could also be `.enabled` or `.restrictedSoon`. While merchant-facing they see no difference, internally makes things confusing and could lead to errors 🤔 

We can see this if we start to test for `.enabled` for example. This would fail because the account state returns `.completed`:

```
    func test_onboarding_returns_enabled_when_PaymentGatewayAccount_status_is_enabled() {
        // Given
        setupCountry(country: .us)
        setupWCPayPlugin(status: .networkActive, version: WCPayPluginVersion.minimumSupportedVersion)
        setupPaymentGatewayAccount(accountType: WCPayAccount.self, status: .enabled)

        // When
        let useCase = CardPresentPaymentsOnboardingUseCase(storageManager: storageManager,
                                                           stores: stores,
                                                           cardPresentPaymentOnboardingStateCache: onboardingStateCache)
        let state = useCase.state

        // Then
        XCTAssertEqual(state, .enabled)
    }
```

Comment (joshheald) [reply Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:12]: I agree that there is potential value in having these states in our onboarding state enum. It might even be ideal to match our onboarding enum to the possible WooPayments states, or have such an enum as a property of our onboarding state.

There are app specific states though:
- select a plugin
- no plugin installed
- IPP not supported in this country/plugin
- more besides...

So the refactor might be quite in depth, and doesn't add a lot of value right now, IMO. It's likely to have most value if we ever add another payment provider to WooPayments.

@jaclync making the state readonly is helpful, thanks for doing that. Perhaps also adding some comment to the Onboarding State enum around the differences between it and the WooPayments onboarding state would be helpful for future engineers touching this?

Approving now, but it would be great if you could add something before merging 😊 

Thanks for discussing and surfacing this complexity

**Review by joshheald — APPROVED** [review]
Approved pending addition of a comment around onboarding state/WooPayments account status differences.

File: Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift
Comment (iamgabrielma) [reply Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:12]: > the refactor might be quite in depth, and doesn't add a lot of value right now, IMO. It's likely to have most value if we ever add another payment provider to WooPayments.

Agree! I quickly logged this separately so it doesn't get lost in GH comments https://github.com/woocommerce/woocommerce-ios/issues/11722

Comment (jaclync) [reply Yosemite/Yosemite/Model/Enums/CardPresentPaymentsOnboardingState.swift:12]: > Perhaps also adding some comment to the Onboarding State enum around the differences between it and the WooPayments onboarding state would be helpful for future engineers touching this?

Yup good idea, added in https://github.com/woocommerce/woocommerce-ios/pull/11690/commits/bc0931ef7be9d72e5a21ea674aed8790d7403cf6.

> At the moment when we perform the account status checks via `CardPresentPaymentsOnboardingUseCase.accountChecks` and find that internally the `accountStatusAllowedForCardPresentPayments` check passes, we're returning that the `CardPresentPaymentOnboardingState` is `.completed` [if does not return an error](https://github.com/woocommerce/woocommerce-ios/blob/trunk/WooCommerce/Classes/ViewRelated/Dashboard/Settings/In-Person%20Payments/CardPresentPaymentsOnboardingUseCase.swift#L407-L417), and while is true that merchants can perform operations at this point as if the onboarding was completed, the account state is not correct at this point, since could also be `.enabled` or `.restrictedSoon`. While merchant-facing they see no difference, internally makes things confusing and could lead to errors 🤔

If the UX is the same for the enabled/restrictedSoon cases, `CardPresentPaymentOnboardingState` (the UI state) staying the same as the completed case is probably fine unless we need to use the specific WooPayments state for other purposes like logging or further handling of each case. Like Josh said https://github.com/woocommerce/woocommerce-ios/pull/11690#discussion_r1455253131, it could be helpful to add the WooPayments state to the UI state. Like adding an associated value of `WCPayAccountStatusEnum` to the `CardPresentPaymentOnboardingState.completed` case.

> I quickly logged this separately so it doesn't get lost in GH comments https://github.com/woocommerce/woocommerce-ios/issues/11722

Thanks for creating the issue!

**Comment by jaclync** [comment]
Additional comments added in https://github.com/woocommerce/woocommerce-ios/pull/11690/commits/bc0931ef7be9d72e5a21ea674aed8790d7403cf6, merging soon for the release before my AFK tomorrow!

**Pull request merged**
